### PR TITLE
⚗️ [RUMF-1337] add `fetch_duration` experiment to resolve incorrect fetch duration

### PR DESCRIPTION
**Pull request by liywjl**
## Motivation

It appears that `.then` is not a good measurement for when a `fetch` request has completed. This means that our own calculate for `fetch` duration have been under reporting the total time of the request.

## Why?

It appears that `fetch(xxx).then(xx)` is triggered when a  response is received. That does not include the response payload being downloaded, which should be included in the duration estimation.

## Solution

We therefore need to instrument the response and listen to it's stream to only calculate duration after the payload has been downloaded.

We should do this behind `fetch_duration` experiment flag.

## Other impacts

- This should increase the likelihood of finding [PerformanceResourceTiming](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceResourceTiming) for `fetch` requests
- Users will no longer be able to use `Response.body.cancel()` or `Response.body.getReader().cancel()` and expect the stream no longer to stop (because we clone the `Response.body` stream)

## Changes

- Refactor and mutualise `readBytesFromStream`
- Add function `waitForResponseToComplete` in `requestCollection` to generate `REQUEST_COMPLETED` after stream is finished
- Replace `context.duration` with `context.resolveDuration` in `fetchObservable`
- add `computeResponseDurationInfo` in `resourceCollection` to get diff details between `resolveDuration` and new `duration` value
- Hide changes behind `fetch_duration` feature flag

## Testing

- [x] Local
- [x] Staging
- [x] Unit
- [x] End to end

---

I have gone over the [contributing](https://github.com/DataDog/browser-sdk/blob/main/CONTRIBUTING.md) documentation.

### CODE REVIEW

File: packages/core/src/browser/fetchObservable.ts [hunk 1]
```diff
@@ -93,25 +94,58 @@ function afterSend(
...
+              return pump()
+            })
+          }
+        },
+      })
+    } else {
+      context.duration = elapsed(context.startClocks.timeStamp, timeStampNow())
+      reportFetch(context, response)
     }
   }
```
Comment (BenoitZugmeyer) [inline packages/core/src/browser/fetchObservable.ts:141]: 💬 **suggestion**: ‏extract this logic in a function that you would use like this:

```ts
waitForResponseToFinish(response).then(() => {
   context.duration = elapsed(context.startClocks.timeStamp, timeStampNow())
   reportFetch(context, response)
})
```


**❓ question:** ‏in a previous PR, I remember you used to factorize this code with the [logs network error collection](https://github.com/DataDog/browser-sdk/blob/0ab07cb70ac32cdd92677e80dd2942e7770e2cb9/packages/logs/src/domain/logsCollection/networkError/networkErrorCollection.ts#L142). Would it possible to do the same here?

File: packages/core/src/browser/fetchObservable.ts [hunk 2]
```diff
@@ -93,25 +94,58 @@ function afterSend(
...
+    const responseClone = response.clone()
+    const reader = responseClone.body?.getReader()
+
+    if (reader && ReadableStream) {
+      new ReadableStream({
+        start(controller) {
+          return pump()
+
+          function pump(): Promise<undefined> {
+            return (reader as ReadableStreamDefaultReader<Uint8Array>).read().then(({ done }) => {
```
Comment (BenoitZugmeyer) [inline packages/core/src/browser/fetchObservable.ts:125]: 💬 **suggestion**: You should handle the case where `.read()` returns a rejected promise (for example, when the request has been cancelled via an `AbortController`). Also `monitor()` those callbacks.

File: test/e2e/scenario/logs.scenario.ts [hunk 3]
```diff
@@ -104,7 +101,7 @@ describe('logs', () => {
         // When reading the request, chunks length are probably not aligning perfectly with the
         // response length limit, so it sends few more bytes than necessary. Add a margin of error
         // to verify that it's still close to the expected limit.
-        DEFAULT_REQUEST_ERROR_RESPONSE_LENGTH_LIMIT * 2
+        DEFAULT_REQUEST_ERROR_RESPONSE_LENGTH_LIMIT * 4
```
Comment (BenoitZugmeyer) [inline test/e2e/scenario/logs.scenario.ts:104]: **🔨 warning:** this assertion is now pointless:

* `getLargeResponseWroteSize` is [around 100K](https://github.com/DataDog/browser-sdk/blob/0ab07cb70ac32cdd92677e80dd2942e7770e2cb9/test/e2e/lib/framework/serverApps/mock.ts#L7)
* `DEFAULT_REQUEST_ERROR_RESPONSE_LENGTH_LIMIT * 4` is [around 128K](https://github.com/DataDog/browser-sdk/blob/0ab07cb70ac32cdd92677e80dd2942e7770e2cb9/packages/logs/src/domain/configuration.ts#L35)

So this test does not ensure that we "read only the first bytes of the response" when doing a fetch.

File: test/e2e/scenario/logs.scenario.ts [hunk 4]
```diff
@@ -78,18 +78,15 @@ describe('logs', () => {
...
-          },
-          (error) => console.log(error)
-        )
+        const controller = new AbortController()
+        const signal = controller.signal
+
+        setTimeout(() => {
+          controller.abort()
+          setTimeout(() => done(undefined), 1000)
+        }, 2000)
```
Comment (BenoitZugmeyer) [inline test/e2e/scenario/logs.scenario.ts:87]: **💭 thought:** ‏those timeout usages seems strange to me, we should talk about it

File: test/e2e/scenario/rum/resources.scenario.ts [hunk 5]
```diff
@@ -196,6 +196,50 @@ describe('rum resources', () => {
       })
   })
 
+  createTest('track redirect fetch timings')
+    .withRum()
+    .run(async ({ serverEvents }) => {
+      await browserExecuteAsync((done) => {
+        fetch(`/redirect?duration=${200}`).then(
```
Comment (BenoitZugmeyer) [inline test/e2e/scenario/rum/resources.scenario.ts:203]: **🥜 nitpick:** ‏
```suggestion
        fetch(`/redirect?duration=200`).then(
```

File: test/e2e/scenario/rum/resources.scenario.ts [hunk 6]
```diff
@@ -196,6 +196,50 @@ describe('rum resources', () => {
...
+      expectToHaveValidTimings(resourceEvent)
+      expect(resourceEvent.resource.redirect).not.toBeUndefined()
+      expect(resourceEvent.resource.redirect!.duration).toBeGreaterThan(0)
+    })
+
+  createTest('track dns lookup fetch timings')
+    .withRum()
+    .run(async ({ serverEvents }) => {
+      await browserExecuteAsync((done) => {
+        fetch(`/redirect?duration=${200}`).then(
```
Comment (BenoitZugmeyer) [inline test/e2e/scenario/rum/resources.scenario.ts:225]: **🥜 nitpick:** ‏
```suggestion
        fetch(`/redirect?duration=200`).then(
```

File: packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts [hunk 7]
```diff
@@ -60,6 +61,14 @@ function processRequest(
   const type = request.type === RequestType.XHR ? ResourceType.XHR : ResourceType.FETCH
 
   const matchingTiming = matchRequestTiming(request)
+  if (!matchingTiming) {
+    addTelemetryDebug('Missing PerformanceResourceTiming', {
+      duration: request.duration,
+      resolveDuration: request.resolveDuration,
+      status: request.status,
+      requestType: request.type === RequestType.XHR ? 'xhr' : 'fetch',
+    })
```
Comment (BenoitZugmeyer) [inline packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts:70]: **💬 suggestion:** ‏this might yield a *lot* of logs. Maybe log it only the first time we got a missing timing?

File: test/e2e/scenario/rum/resources.scenario.ts [hunk 8]
```diff
@@ -196,6 +196,50 @@ describe('rum resources', () => {
...
+      expect(resourceEvent).not.toBeUndefined()
+      expect(resourceEvent.resource.method).toEqual('GET')
+      expect(resourceEvent.resource.status_code).toEqual(200)
+      expect(resourceEvent._dd.resolveDuration).toBeGreaterThan(0)
+      expectToHaveValidTimings(resourceEvent)
+      expect(resourceEvent.resource.redirect).not.toBeUndefined()
+      expect(resourceEvent.resource.redirect!.duration).toBeGreaterThan(0)
+    })
+
+  createTest('track dns lookup fetch timings')
```
Comment (BenoitZugmeyer) [inline test/e2e/scenario/rum/resources.scenario.ts:221]: **❓ question:** ‏ ‏this test is an exact copy of the one before. Typo?

File: packages/core/src/browser/fetchObservable.ts (hunk 2 above)
Comment (liywjl) [reply packages/core/src/browser/fetchObservable.ts:125]: we already use `monitor` on `responsePromise.then(monitor(onFulfilled), monitor(onRejected))`, does it make sense to have it here too?

Comment (BenoitZugmeyer) [reply packages/core/src/browser/fetchObservable.ts:125]: Ah that's right the promise is handled somewhere else. Nevermind my comment then!

File: test/e2e/scenario/logs.scenario.ts (hunk 4 above)
Comment (liywjl) [reply test/e2e/scenario/logs.scenario.ts:87]: This wasn't ideal I agree. I was trying to call `controller.abort()` and `done()` on `.then` but found that no entry would be present in `serverEvents`.

This workaround gets around the issue. But I agree it isnt ideal

File: packages/core/src/browser/fetchObservable.ts [hunk 9]
```diff
@@ -93,25 +94,63 @@ function afterSend(
...
+
+  responsePromise.then(
+    monitor((response) =>
+      waitForResponseToFinish(response).then(
+        (resolveDuration: Duration) => reportFetch(response, resolveDuration),
+        () => reportFetch(response)
+      )
+    ),
+    monitor(reportFetch)
+  )
```
Comment (bcaudan) [inline packages/core/src/browser/fetchObservable.ts:155]: **💭 thought:** Since we are already reading responses for logs network error collection, there are probably some thoughts to have on that front.

- Do we really need to wait for response to finish for all fetch instrumentation usages? 
- Should we mutualise some things with our current [response reading code](https://github.com/DataDog/browser-sdk/blob/main/packages/logs/src/domain/logsCollection/networkError/networkErrorCollection.ts#L201)?

File: test/e2e/scenario/logs.scenario.ts (hunk 4 above)
Comment (bcaudan) [inline test/e2e/scenario/logs.scenario.ts:87]: **💭 thought:** ‏Previous comment was useful to understand why we were canceling the response.
No idea what was the issue with your changes and what you are trying to do.
Could you clarify that?

File: packages/rum-core/src/rawRumEvent.types.ts [hunk 10]
```diff
@@ -41,6 +41,7 @@ export interface RawRumResourceEvent {
     span_id?: string // not available for initial document tracing
     rule_psr?: number
     discarded: boolean
+    resolveDuration?: ServerDuration
```
Comment (bcaudan) [inline packages/rum-core/src/rawRumEvent.types.ts:44]: **❓ question:** ‏is it only for while we are monitoring the change or do we plan to keep this property?

File: packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts [hunk 11]
```diff
@@ -60,6 +63,17 @@ function processRequest(
   const type = request.type === RequestType.XHR ? ResourceType.XHR : ResourceType.FETCH
 
   const matchingTiming = matchRequestTiming(request)
+
+  if (!hasLoggedMissingResource && !matchingTiming) {
+    addTelemetryDebug('Missing PerformanceResourceTiming', {
```
Comment (bcaudan) [inline packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts:68]: **❓ question:** ‏matchTiming can be undefined for several reasons, what do you want to monitor exactly?‏

File: packages/core/src/browser/fetchObservable.ts (hunk 9 above)
Comment (liywjl) [reply packages/core/src/browser/fetchObservable.ts:155]: > Should we mutualise some things with our current response reading code?
Yes agreed

> Do we really need to wait for response to finish for all fetch instrumentation usages?
I agree this isn't ideal, however if we wanted to finish earlier we cannot determine a precise duration. Happy to sync up about this issue

File: packages/core/src/browser/fetchObservable.ts [hunk 12]
```diff
@@ -96,22 +96,18 @@ function afterSend(
   const reportFetch = (response: Response | Error) => {
     const context = startContext as unknown as FetchCompleteContext
     context.state = 'complete'
-    context.duration = elapsed(context.startClocks.timeStamp, timeStampNow())
```
Comment (liywjl) [inline packages/core/src/browser/fetchObservable.ts:99]: We know duration is incorrect here so it makes sense for us to remove it

File: packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts [hunk 13]
```diff
@@ -52,6 +60,22 @@ export function startResourceCollection(
   })
 }
 
+function waitForResponseToFinish(request: RequestCompleteEvent, callback: (duration: Duration) => void) {
+  const duration = request.duration || elapsed(request.startClocks.timeStamp, timeStampNow())
```
Comment (liywjl) [inline packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts:64]: for `XHR` request.duration should be defined, else we should create one as a fallback

File: test/e2e/scenario/logs.scenario.ts [hunk 14]
```diff
@@ -74,22 +74,15 @@ describe('logs', () => {
...
-        fetch('/throw-large-response').then(
-          (response) => {
-            // The body stream needs to be cancelled, else the browser will still download the whole
-            // response even if it is unused.
-            response
-              .body!.getReader()
-              .cancel()
-              .catch((error) => console.log(error))
+        fetch('/throw-large-response')
+          .then((response: Response) => response.clone().blob())
```
Comment (BenoitZugmeyer) [inline test/e2e/scenario/logs.scenario.ts:82]: **🔨 warning:** ‏the `.clone()` here looks fishy. The user shouldn't need to clone their response... let's see why we need it

File: packages/core/src/tools/readBytesFromStream.ts [hunk 15]
```diff
@@ -0,0 +1,75 @@
+import { monitor } from './monitor'
+import { noop } from './utils'
+
+type Options = {
+  limit: number
```
Comment (amortemousque) [inline packages/core/src/tools/readBytesFromStream.ts:5]: **🥜 nitpick:** ‏ ‏Maybe the naming could be more explicit like `bytesLimit`

File: packages/rum-core/src/domain/requestCollection.ts [hunk 16]
```diff
@@ -39,6 +39,7 @@ export interface RequestCompleteEvent {
   status: number
   responseType?: string
   startClocks: ClocksState
+  resolveDuration?: Duration
   duration: Duration
```
Comment (amortemousque) [inline packages/rum-core/src/domain/requestCollection.ts:43]: **❓ question:** ‏Should it be optional?
```suggestion
  duration?: Duration
```

File: packages/core/src/browser/fetchObservable.ts [hunk 17]
```diff
@@ -96,22 +97,19 @@ function afterSend(
...
-
+    context.resolveDuration = elapsed(startContext.startClocks.timeStamp, timeStampNow())
```
Comment (amortemousque) [inline packages/core/src/browser/fetchObservable.ts:100]: **❓ question:** ‏Shouldn't it be in `FetchCompleteContext` instead of the `duration` attribute?

File: packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts [hunk 18]
```diff
@@ -37,7 +41,12 @@ export function startResourceCollection(
   sessionManager: RumSessionManager
 ) {
   lifeCycle.subscribe(LifeCycleEventType.REQUEST_COMPLETED, (request: RequestCompleteEvent) => {
-    lifeCycle.notify(LifeCycleEventType.RAW_RUM_EVENT_COLLECTED, processRequest(request, configuration, sessionManager))
+    waitForResponseToFinish(request, () =>
```
Comment (amortemousque) [inline packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts:44]: **❓ question:** ‏Why not waiting for response to finish inside the `requestCollection`?

File: packages/core/src/browser/fetchObservable.ts (hunk 17 above)
Comment (liywjl) [reply packages/core/src/browser/fetchObservable.ts:100]: I am not sure I follow, do you mean to say that the variable should probably be defined in `beforeSend` where we define the `FetchCompleteContext`?

If so, I do not think this is the right place to put it. 

`reportFetch` gets called only after the `fetch` `.then` has been triggered. Moving this logic to `FetchCompleteContext` (aka `beforeSend`) would mean we incorrectly calculating `resolveDuration`

File: packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts (hunk 18 above)
Comment (liywjl) [reply packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts:44]: Good point, I initially put it here as this was where `PerformanceEntry` logic resided. But bar that there was no real reason for it to be here.

It probably is better for us to move this logic into `requestCollection` as it is closer to the instrumentation and thus will give us a better `duration` estimation

File: packages/rum-core/src/domain/requestCollection.ts (hunk 16 above)
Comment (liywjl) [reply packages/rum-core/src/domain/requestCollection.ts:43]: Good point, I will extend this type to make sure we reflect when `RequestCompleteEvent` contains and does not contain duration

File: packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts [hunk 19]
```diff
@@ -66,6 +102,13 @@ function processRequest(
   const tracingInfo = computeRequestTracingInfo(request, configuration)
   const indexingInfo = computeIndexingInfo(sessionManager, startClocks)
 
+  const durationDiff = request.resolveDuration
+    ? toServerDuration((request.duration - request.resolveDuration) as Duration)
+    : undefined
+  const durationPercentageDiff = durationDiff
+    ? Math.round((durationDiff / toServerDuration(request.duration)) * 100)
+    : undefined
+
```
Comment (BenoitZugmeyer) [inline packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts:111]: **💭 thought:** ‏I'm not sure we need those, as we can do arithmetics with measurement facets directly in Datadog


**💬 suggestion:** ‏if we really need those, I would do something like this:
```ts
let durationDetails
if (request.resolveDuration) {
  durationDetails = {
    resolveDuration: ...,
    durationDiff: ...,
    durationPercentageDiff: ...
  }
}
```

File: test/e2e/lib/framework/createTest.ts [hunk 20]
```diff
@@ -20,7 +20,7 @@ const DEFAULT_RUM_CONFIGURATION = {
   trackLongTasks: true,
   telemetrySampleRate: 100,
   telemetryConfigurationSampleRate: 100,
-  enableExperimentalFeatures: [],
+  enableExperimentalFeatures: ['fetch_duration'],
```
Comment (BenoitZugmeyer) [inline test/e2e/lib/framework/createTest.ts:23]: **💬 suggestion:** ‏I'm not sure we want to enable it for all tests, as we want to ensure that the SDK works without this feature too.
* If you added a test that requires this flag to work, you can use `withRum({ enableExperimentalFeatures: ['fetch_duration'] })`
* if you want to make sure an existing test still works with this flag, you can duplicate it and use `withRum({ enableExperimentalFeatures: ['fetch_duration'] })` on one of them.

File: packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts (hunk 19 above)
Comment (liywjl) [reply packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts:111]: > we can do arithmetics with measurement facets

This was done based on @bcaudan 's suggestion. Happy to remove it if you insist, I personally think it's fine and do plan on deleting this section in a future PR, so I do not have super strong feelings either way.

> I would do something like ...

I agree its not very nice looking. I will add `nullish coalescing` and the `if` statement to make it a little nicer. I do plan on removing this section in a future PR shortly

File: packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts (hunk 18 above)
Comment (amortemousque) [reply packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts]: If we put it in `requestCollection` and wait For Response To Finish  before triggering `REQUEST_COMPLETED` it will also affect `waitPageActivityEnd`. Conceptually it would make sens to me but might have broader impact. @BenoitZugmeyer wdyt?

Comment (BenoitZugmeyer) [reply packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts]: Makes sense to me too! The UI usually react when the response finishes to load

**Review by BenoitZugmeyer — APPROVED** [review]

File: packages/rum-core/src/domain/requestCollection.ts [hunk 21]
```diff
@@ -119,23 +127,25 @@ export function trackFetch(lifeCycle: LifeCycle, configuration: RumConfiguration
...
-          startClocks: context.startClocks,
-          status: context.status,
-          traceId: context.traceId,
-          traceSampled: context.traceSampled,
-          type: RequestType.FETCH,
-          url: context.url,
-          response: context.response,
-          init: context.init,
-          input: context.input,
+        waitForResponseToFinish(context, (duration: Duration) => {
```
Comment (bcaudan) [inline packages/rum-core/src/domain/requestCollection.ts:130]: **🥜 nitpick:** ‏We could maybe make clearer the lifecycle difference with xhr by renaming `complete` occurrences for fetch into something like `resolve`, wdyt?

File: packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts [hunk 22]
```diff
@@ -66,6 +66,13 @@ function processRequest(
   const tracingInfo = computeRequestTracingInfo(request, configuration)
   const indexingInfo = computeIndexingInfo(sessionManager, startClocks)
 
+  let durationDiff
+  let durationPercentageDiff
+  if (request.resolveDuration) {
+    durationDiff = toServerDuration((request.duration - request.resolveDuration) as Duration)
+    durationPercentageDiff = Math.round((durationDiff / toServerDuration(request.duration)) * 100)
+  }
```
Comment (bcaudan) [inline packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts:74]: **🥜 nitpick:** ‏what about extracting all these infos into a `computeXXXInfo` like other parts of this function?

File: packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.spec.ts [hunk 23]
```diff
@@ -135,6 +139,9 @@ describe('resourceCollection', () => {
       type: RumEventType.RESOURCE,
       _dd: {
         discarded: false,
+        resolveDuration: (50 * 1e6) as ServerDuration,
+        durationDiff: 50000000 as ServerDuration,
```
Comment (bcaudan) [inline packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.spec.ts:143]: **💬 suggestion:** ‏for consistency sake
```suggestion
        resolveDuration: (50 * 1e6) as ServerDuration,
        durationDiff: (50 * 1e6) as ServerDuration,
```

File: packages/core/src/tools/readBytesFromStream.spec.ts [hunk 24]
```diff
@@ -0,0 +1,93 @@
+import { isIE } from './browserDetection'
+import { readBytesFromStream } from './readBytesFromStream'
+
+describe('readBytesFromStream', () => {
```
Comment (bcaudan) [inline packages/core/src/tools/readBytesFromStream.spec.ts:4]: **💭 thought:** ‏some tests in `networkErrorCollection.spec` could be more suited here now

File: test/e2e/scenario/logs.scenario.ts [hunk 25]
```diff
@@ -74,22 +74,15 @@ describe('logs', () => {
...
-        fetch('/throw-large-response').then(
-          (response) => {
-            // The body stream needs to be cancelled, else the browser will still download the whole
-            // response even if it is unused.
-            response
-              .body!.getReader()
-              .cancel()
-              .catch((error) => console.log(error))
+        fetch('/throw-large-response')
+          .then((response: Response) => response.blob())
```
Comment (bcaudan) [inline test/e2e/scenario/logs.scenario.ts:82]: **💬 suggestion:** ‏adding a comment to explain what is your intent with `response.blob()` could be helpful.

File: test/e2e/scenario/rum/resources.scenario.ts [hunk 26]
```diff
@@ -194,8 +194,47 @@ describe('rum resources', () => {
...
+
+          fetch('/large-response', { signal }).then(() => {
+            controller.abort()
+            done(undefined)
+          }, console.log)
+        })
+
+        await flushEvents()
+        expect(servers.base.app.getLargeResponseWroteSize()).toBeLessThan(LARGE_RESPONSE_MIN_BYTE_SIZE)
+      })
```
Comment (bcaudan) [inline test/e2e/scenario/rum/resources.scenario.ts:214]: **❓ question:** is this behavior related to SDK code?

File: packages/rum-core/src/domain/requestCollection.ts [hunk 27]
```diff
@@ -148,3 +158,23 @@ function getNextRequestIndex() {
...
+        },
+        {
+          bytesLimit: Number.POSITIVE_INFINITY,
+          collectStreamBody: false,
+        }
+      )
+      return
+    }
+  }
+  callback(elapsed(context.startClocks.timeStamp, timeStampNow()))
```
Comment (bcaudan) [inline packages/rum-core/src/domain/requestCollection.ts:179]: **💭 thought:** ‏would it worth having tests for null response/body behaviors?

File: packages/core/src/tools/readBytesFromStream.spec.ts (hunk 24 above)
Comment (bcaudan) [reply packages/core/src/tools/readBytesFromStream.spec.ts:4]: Unit tests in `networkErrorCollection.spec` should be specific to network collection logic and tests in `readBytesFromStream.spec` should be specific to byte read logic.
By having a quick look `does not yield an unhandled rejection error if the cancel promise is rejected` seems to be misplaced to me. is it the case? and if yes, could other ones be misplaced as well?

Comment (liywjl) [reply packages/core/src/tools/readBytesFromStream.spec.ts:4]: All tests under `describe('computeFetchResponseText'` focus on testing logic from `readBytesFromStream` (especially as we do not test our fallback cases). 

Are you suggesting we remove all of them and add the missing ones to `readBytesFromStream.spec`?

Comment (bcaudan) [reply packages/core/src/tools/readBytesFromStream.spec.ts:4]: I'd like to keep the same level of coverage but have the tests as close as possible of the logic they are testing.
Since you extracted the logic, you should extract and adjust the corresponding tests as well.
Does it make sense?

File: test/e2e/scenario/logs.scenario.ts [hunk 28]
```diff
@@ -78,6 +78,8 @@ describe('logs', () => {
     .withLogs({ forwardErrorsToLogs: true })
     .run(async ({ serverEvents, baseUrl, servers }) => {
       await browserExecuteAsync((done) => {
+        // This example formerly used "response.body!.getReader().cancel()" to cancel the stream midway.
+        // However, as we not longer support ".cancel()" in streams and thus why we use "response.blob()".
```
Comment (bcaudan) [inline test/e2e/scenario/logs.scenario.ts:82]: **❓ question:** ‏
does using blob cancel the stream? 
if not, why are we using `response.blob` and not calling `done` directly after the fetch resolve?
any reason to use `blob` and not `json` or `text` instead?

File: packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts [hunk 29]
```diff
@@ -183,6 +178,19 @@ function computeEntryTracingInfo(entry: RumPerformanceResourceTiming, configurat
...
+  return {
+    durationDiff,
+    durationPercentageDiff,
+  }
```
Comment (bcaudan) [inline packages/rum-core/src/domain/rumEventsCollection/resource/resourceCollection.ts:191]: **💬 suggestion:** ‏what about encapsulating all the response duration info logic/format in this method and return directly:
```ts
return {
  _dd: {
    resolveDuration,
    durationDiff,
    durationPercentageDiff
  }
}
```
it is what we are doing on other methods

File: packages/rum-core/src/domain/requestCollection.ts [hunk 30]
```diff
@@ -159,7 +159,7 @@ function getNextRequestIndex() {
   return result
 }
 
-function waitForResponseToFinish(context: RumFetchCompleteContext, callback: (duration: Duration) => void) {
+function waitForResponseToFinish(context: RumFetchResolveContext, callback: (duration: Duration) => void) {
```
Comment (bcaudan) [inline packages/rum-core/src/domain/requestCollection.ts:162]: **💬 suggestion:** `waitForResponseToComplete` would avoid to use two terms for the same concept‏

File: packages/logs/src/domain/logsCollection/networkError/networkErrorCollection.ts [hunk 31]
```diff
@@ -27,12 +27,12 @@ export function startNetworkErrorCollection(configuration: LogsConfiguration, li
...
   })
   const fetchSubscription = initFetchObservable().subscribe((context) => {
-    if (context.state === 'complete') {
+    if (context.state === 'resolve') {
       handleCompleteRequest(RequestType.FETCH, context)
     }
   })
 
-  function handleCompleteRequest(type: RequestType, request: XhrCompleteContext | FetchCompleteContext) {
+  function handleCompleteRequest(type: RequestType, request: XhrCompleteContext | FetchResolveContext) {
```
Comment (bcaudan) [inline packages/logs/src/domain/logsCollection/networkError/networkErrorCollection.ts:35]: **💬 suggestion:** ‏`handleCompleteRequest` is not accurate anymore, maybe something like `handleResponse` would be enough

File: packages/rum-core/src/domain/requestCollection.spec.ts [hunk 32]
```diff
@@ -100,6 +100,33 @@ describe('collect fetch', () => {
     })
   })
 
+  describe('with feature flag fetch_duration', () => {
+    beforeEach(() => updateExperimentalFeatures(['fetch_duration']))
+    afterEach(() => resetExperimentalFeatures())
+
+    it('should still function if response is defined', (done) => {
```
Comment (bcaudan) [inline packages/rum-core/src/domain/requestCollection.spec.ts:107]: **💬 suggestion:** ‏`still function` is quite vague, looking at other tests something like `should notify when response is defined` would be more consistent

File: test/e2e/scenario/logs.scenario.ts (hunk 28 above)
Comment (liywjl) [reply test/e2e/scenario/logs.scenario.ts:82]: > does using blob cancel the stream?

Good point, no. I will update the comment.

> why are we using response.blob and not calling done directly after the fetch resolve?

I initially had issues getting this condition to work because we were aborting the call too early to be able to fetch the first couple bytes.

I assumed calling `done` would also abort the stream too early for us to collect the bytes, but that does not seem to be the case. I will therefore not use any response methods (`blob`,`json` and `text`) and simply use `done`

> any reason to use blob and not json or text instead?

No, we do not care about the contents of the response (also we are not going to use any of these methods).

**Review by bcaudan — APPROVED** [review]
